 <div style="border-radius: 15px; box-shadow: 2px 2px 4px; border: 1px solid; background:#bdc3c7; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<h7 style="color:black; margin-bottom:20px"> Людмила, привет!

Мои комментарии и замечания далее по тексту помечены различными цветами:

---
<div class="alert alert-block alert-danger">
⁉️ : Критическое замечание, которые следует исправить.
</div>

<div class="alert alert-block alert-warning">
✏️ : Рекомендация\совет на будущее.
</div>

<div class="alert alert-block alert-success">
✔️ : Когда всё сделано правильно.
</div>

---

Хорошей практикой является оставлять свои комментарии после исправлений замечаний или для вопросов, разъяснений. Для того чтобы мне было легче их найти, то выделяй, пожалуйста, с помощью цвета, например вот так:

<div class="alert alert-info" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " ><b> Комментарий студента : </b>

Вопрос, который не останется незамеченным

</div>

---

**В рамках финального спринта у вас есть лимит на количество итераций проверок. Я буду для информативности их отмечать в итоговом ревью:**

<div class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid ">

**`[1/6]`**

Итог ревью - ...

</div>

p.s.: не удаляй мои замечания, если предстоит что-то доработать в проекте.</h7>

---

<div id="1" class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<b> ⚠️ ИТОГИ РЕВЬЮ ⚠️ : </b>

**`[1/6]`**

<s>Большинство запросов выполнены корректно и возвращают правильные результаты. Однако есть одно задание, которое требует доработки:

 - [Ошибка №1](#div_id1)

Пожалуйста, обрати внимание и внеси соответствующие изменения.

</div>

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<b> 🚀 ИТОГИ РЕВЬЮ 🚀 : </b>

Все замечания учтены. Эта работа демонстрирует твоё умение формировать запросы SQL для решения подобных задач.

Поздравляю с успешным завершением финального этапа. Особенно ценно, что все проекты выполнены в установленные сроки 👏

---

![](https://media.giphy.com/media/9hLq3XV0tykOINfyXc/giphy.gif)

---

После окончания дедлайнов представлю в канале Пачки дополнительные материалы для дальнейшего развития в этой области.

Успехов тебе в дальнейших начинаниях!

</div>

---

# Проект "Анализ базы данных SQL"

**Описание проекта:**
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

**Наша первая задача** — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель исследования:** 
Проанализировать базу данных для разработки ценностного предложения для нового продукта

**Описание даных:**

Таблица **books** (содержит данные о книгах):
 - `book_id` — идентификатор книги;
 - `author_id` — идентификатор автора;
 - `title` — название книги;
 - `num_pages` — количество страниц;
 - `publication_date` — дата публикации книги;
 - `publisher_id` — идентификатор издателя.

Таблица **authors** (содержит данные об авторах):
 - `author_id` — идентификатор автора;
 - `author` — имя автора.

Таблица **publishers** (содержит данные об издательствах):
 - `publisher_id` — идентификатор издательства;
 - `publisher` — название издательства.

Таблица **ratings** (содержит данные о пользовательских оценках книг):
 - `rating_id` — идентификатор оценки;
 - `book_id` — идентификатор книги;
  - `username` — имя пользователя, оставившего оценку;
 - `rating` — оценка книги.

Таблица **reviews** (содержит данные о пользовательских обзорах на книги):
 - `review_id` — идентификатор обзора;
 - `book_id` — идентификатор книги;
 - `username` — имя пользователя, написавшего обзор;
 - `text` — текст обзора.


## Импорт библиотек. Установка параметров подключения. Функция запроса SQL

In [1]:
# импортирт библиотек
import pandas as pd
import sqlalchemy as sa

In [2]:
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)

## Исследование таблиц базы данных

### Таблица books (содержит данные о книгах)

In [4]:
# Посмотрим на первые 5 строк таблицы
query = '''
SELECT *
FROM books
LIMIT 5
'''
get_sql_data(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
# Подсчитаем общее количество книг
query = 'SELECT COUNT(*) FROM books';
get_sql_data(query)

count
0   1000

### Таблица authors (содержит данные о книгах)

In [6]:
# Посмотрим на первые 5 строк таблицы
query = '''
SELECT *
FROM authors
LIMIT 5
'''
get_sql_data(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
# Подсчитаем общее количество строк
query = 'SELECT COUNT(*) FROM authors';
get_sql_data(query)

count
0    636

### Таблица publishers (содержит данные об издательствах)

In [8]:
# Посмотрим на первые 5 строк таблицы
query = '''
SELECT *
FROM publishers
LIMIT 5
'''
get_sql_data(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
# Подсчитаем общее количество строк
query = 'SELECT COUNT(*) FROM publishers';
get_sql_data(query)

count
0    340

### Таблица ratings (содержит данные о пользовательских оценках книг)

In [10]:
# Посмотрим на первые 5 строк таблицы
query = '''
SELECT *
FROM ratings
LIMIT 5
'''
get_sql_data(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
# Подсчитаем общее количество строк
query = 'SELECT COUNT(*) FROM ratings';
get_sql_data(query)

count
0   6456

### Таблица reviews (содержит данные о пользовательских обзорах на книги)

In [12]:
# Посмотрим на первые 5 строк таблицы
query = '''
SELECT *
FROM reviews
LIMIT 5
'''
get_sql_data(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
# Подсчитаем общее количество строк
query = 'SELECT COUNT(*) FROM reviews';
get_sql_data(query)

count
0   2793

### Выводы

В процессе знакомства с данными, мы выяснили, что в базе данных находятся сведения о 1000 книг. 

В базе данных представлено 636 авторов и 340 издательств.

Кроме того, есть данные, по отзывам пользователей на книги. Всего мы располагаем 2793 отзывами.
Кроме этого есть данные о 6456 пользовательских оценках.

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<b> ✔️ Комментарий тимлида : </b>

Данные успешно загружены и проанализированы.

Теперь всё готово для дальнейшего выполнения заданий.

</div>

## Задания проекта

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [14]:
query = '''
SELECT 
    COUNT(*) AS cnt
FROM 
    books
WHERE
    publication_date > '2000-01-01'
'''
get_sql_data(query)

cnt
0  819

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<b> ✔️ Комментарий тимлида : </b>

Подход корректный и ответ верный

</div>

**Вывод.**
Всего, после 1 января 2000 года вышло 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [15]:
query = '''
SELECT
    books.title AS book_name,
    subquery_reviews.reviews_qty AS reviews_qty,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    LEFT JOIN (
        SELECT
            book_id,
            COUNT(username) AS reviews_qty
        FROM
            reviews
        GROUP BY
            book_id
        ) AS subquery_reviews ON subquery_reviews.book_id = books.book_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
    books.title,
    subquery_reviews.reviews_qty
ORDER BY
    avg_rating DESC
'''
get_sql_data(query)

book_name  reviews_qty  \
0                                  Homage to Catalonia          2.0   
1                                                March          2.0   
2    In the Hand of the Goddess (Song of the Liones...          2.0   
3                 Welcome to Temptation (Dempseys  #1)          2.0   
4    Stone of Farewell (Memory  Sorrow  and Thorn  #2)          1.0   
..                                                 ...          ...   
995  The World Is Flat: A Brief History of the Twen...          3.0   
996                                              Junky          2.0   
997                  His Excellency: George Washington          2.0   
998                                      Drowning Ruth          3.0   
999                               Harvesting the Heart          2.0   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<b> ✔️ Комментарий тимлида : </b>

Этот запрос считается одним из самых сложных в этом проекте, но результат получен правильно

</div>

**Вывод:**
Посмотрели рейтинги и количество обзоров для каждой книги, обзоров немного.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [16]:
query = ''' WITH 
            search_over50 AS (SELECT book_id, publisher_id FROM books WHERE num_pages > 50)
            SELECT pb.publisher, 
                   COUNT(*) AS cnt
            FROM search_over50 AS s50 
            INNER JOIN  publishers AS pb ON s50.publisher_id=pb.publisher_id
            GROUP BY 
                pb.publisher
            ORDER BY 
                cnt DESC
            LIMIT 1;
        '''
get_sql_data(query)

publisher  cnt
0  Penguin Books   42

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<b> ✔️ Комментарий тимлида : </b>

Согласен с результатом. Действительно, выбранное издательство является лидером по числу выпущенных книг.

</div>

**Вывод:**
Наибольшее количество книг в нашей библиотеке было выпущено издательством "Penguin Books" - 42 книги.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [20]:
query = '''
SELECT
    authors.author,
    AVG(subquery_ratings.avg_rating) AS average_rating
FROM
    books
    LEFT JOIN (
        SELECT 
            book_id,
            AVG(rating) AS avg_rating,
            COUNT(rating) AS rating_qty
        FROM
            ratings
        GROUP BY
            book_id
    ) AS subquery_ratings ON subquery_ratings.book_id = books.book_id
    LEFT JOIN authors ON authors.author_id = books.author_id
WHERE
    rating_qty >= 50
GROUP BY
    author
ORDER BY
    average_rating DESC
LIMIT 1
'''
get_sql_data(query)

author  average_rating
0  J.K. Rowling/Mary GrandPré        4.283844

<div id="div_id1" class="alert alert-danger" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<b> ⁉️ Критическое замечание тимлида : </b>

<s>[⬆ Ошибка №1](#1)


Ты используешь `WHERE rating_qty > 50`, когда по условию задачи, это должно быть нестрогое неравенство `>=`.

Сейчас на результат это не повлияет, но в будущем - по другим задачам такие нюансы могут првиести к неверным результатам


</div>

<div class="alert alert-info" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " ><b> Комментарий студента : </b>

Ой, как невнимательно я. Спасибо!

</div>


**Вывод:**
Самая высокая средняя оценка книг у Д. Роулинг. Средняя оценка ее книг в нашем датасете - 4.28 балла.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [18]:
query = '''
WITH user_ratings AS (
                    SELECT rt.username, 
                           COUNT(rt.rating_id) AS rating_count
                    FROM ratings AS rt
                    GROUP BY rt.username
                    HAVING COUNT(rt.rating_id) > 48
                ),
user_reviews AS (
                SELECT rw.username, 
                       COUNT(rw.review_id) AS review_count
                FROM reviews AS rw
                WHERE rw.username IN (SELECT username FROM user_ratings)
                GROUP BY rw.username
            )
SELECT AVG(review_count) AS avg_reviews
FROM user_reviews;
'''
get_sql_data(query)

avg_reviews
0         24.0

<div class="alert alert-success" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<b> ✔️ Комментарий тимлида : </b>

Это верный ответ. Пользователи корректно отобраны, а их показатели правильно рассчитаны.

</div>

**Вывод:** Пользователи, которые поставили больше 48 оценок, в среднем пишут 24 обзора.

## Общий вывод

1. В процессе знакомства с данными, мы выяснили, что в базе данных находятся сведения о **1000 книг.** 
В базе данных представлено **636 авторов** и **340 издательств**.
Кроме того, есть данные, по отзывам пользователей на книги. Всего мы располагаем **2793 отзывами**.
Кроме этого есть данные о **6456 пользовательских оценках**.

2. Всего, после 1 января 2000 года вышло **819 книг**.
Посмотрели **рейтинги и количество обзоров** для каждой книги, обзоров немного.
Наибольшее количество книг в нашей библиотеке было выпущено издательством **"Penguin Books" - 42 книги**.
Самая высокая средняя оценка книг у **Д. Роулинг. Средняя оценка ее книг в нашем датасете - 4.28 балла**.
Пользователи, которые поставили больше 48 оценок, в среднем пишут **24 обзора**.

**Рекомендации:**
Для клиентов важно найти информацию о книге на сайте, особенно, если они ее не читали. Чем больше будет рейтингов и ревью, тем интереснее, будет ощущение не только ресурса по скачиванию книг, а целое комьюнити. Я бы рекомендовала поощрять за оценки и отзывы о книгах, предоставляя скидки на следующие покупки.

<div class="alert alert-warning" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid; position: relative; padding: 15px; padding-left: 80px;">
<img src="https://lh3.googleusercontent.com/a/AAcHTtc31jlywy-FMYxLAouSSqNAmm5NsrBvivwRIASNt7wihI2ClYdpAbzWNqcigfcR04LfvopHwsJrJpnVr4UXDkCHtWk=s432-c-no" width="45" height="45" style="position: absolute; top: 15px; left: 15px; border-radius: 50%;">
<b> ✏️ Рекомендация тимлида : </b>

Для полноты картины и большей практической ценности результатов, можно было добавить некоторые рекомендации для бизнеса.

</div>

<div class="alert alert-info" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " ><b> Комментарий студента : </b>

Дописала. Спасибо!

</div>
